### DISCUSSION TITLE
channel.flush returning an empty array ?

### DISCUSSION THREAD

**Discussion by challet** [redux-saga/redux-saga#2700] [Q&A] [answered]
Hello, using a channel to buffer messages, I'm taking them out with the `flush` method in order to make some de-duplication computing before to handle them to an other part of the code.

I'm using it like I would with the `take` method, from the example https://redux-saga.js.org/docs/advanced/Channels#using-the-actionchannel-effect.

```ts
const channel = yield actionChannel((action) => actionCreator.match(action) && action.meta.someBoolFlag);
while(true) {
    const items = yield flush(channel);
    console.log(items);
}
```

But it doesn't seem to block until new items are coming in like `take` would, even it is marked as [blocking](https://redux-saga.js.org/docs/api#blocking--non-blocking). So the loop goes infinite with empty arrays. 

Am I getting it wrong about how `flush` works ?

**Comment by Andarist** [comment]
`flush` shouldn't be marked as blocking. We'd appreciate a PR to fix this in the docs.

**Reply by challet** [reply]
Done in #2701 

I suppose I would then need to use `take` in the loop, combined with a `race` effect

**Reply by challet** [reply]
Something like that 

```ts
while(true) {
  const items = [];
  let item;
  let delayed = false;
  do {
    [item, delayed = false] = yield race([
      take(channel),
      delay(10)
    ]);
    if(!delayed) {
      items.push(item);
    }
  } while(!delayed);
  console.log(items);
}
```

**Reply by challet** [reply, marked as answer]
Simpler version with using the [Buffer](https://redux-saga.js.org/docs/api#buffer) api : 

```ts
const buffer = buffers.expanding<ActionType>();
const channel = yield actionChannel((action) => actionCreator.match(action) && action.meta.someBoolFlag);

while(true) {
  const items: ActionType[] = [];
  do {
    items.push(yield take(channel));
  } while(!buffer.isEmpty());

  // do somehting with the items batch
}
```
